src/cascade/input_data/configuration/construct_country.py

```diff
@@ -217,3 +226,187 @@ def convert_gbd_ids_to_dismod_values(with_ids_df, age_groups_df):
     reordered["time_upper"] = reordered["year_id"] + 1
     dropped = reordered.drop(columns=["age_group_id", "year_id", "original_index"])
     return dropped.rename(columns={"age_group_years_start": "age_lower", "age_group_years_end": "age_upper"})
+
+
+def compute_covariate_age_time_dimensions(covariates):
+    """
+    Determines if the input covariate data is by_age and/or by_time.
+
+    Returns:
+        dictionary: keys indicate if age or time are points (0d) vs. 1d
+    """
+
+    covar_at_dims = {}
+
+    covar_at_dims["age_1d"] = len(set(zip(covariates["age_lower"], covariates["age_upper"]))) > 1
+
+    covar_at_dims["time_1d"] = len(set(zip(covariates["time_lower"], covariates["time_upper"]))) > 1
+
+    return covar_at_dims
+
+
+def compute_covariate_age_interval(covariates):
+    """
+    Create an interval expressing all the ages in the covariates data frame.
+    If the covariates have age ranges like: [5,10], [10,15], [30,35], [35-80],
+    the overall age interval would be: [5,15], [30,80]
+    This allows checks like: 9.5 in age_interval (yes), 20.25 in age_interval (no),
+    0.667 in age_interval (no), 95 in age_interval (no), 50.1 in age_interval (yes)
+
+    Returns:
+        Interval: of ages in the covariates data frame
+    """
+    age_interval = it.empty()
+
+    age_groups = set(zip(covariates["age_lower"], covariates["age_upper"]))
+
+    for age in age_groups:
```

Comment:
Instead of forcing the dataframe into a set above I would use `.drop_duplicates` there and `.iterrows` here. For small dataframes they're equivalent but for larger ones staying in pandas makes fewer copies which can be meaningful.

Reply:
ok, did this!